File: DPCodes/452-longestIncreasingPath.py
"""
 a two-dimensional integer matrix, find the length of the longest strictly increasing path. You can move up, down, left, or right.

Constraints1
n, m ≤ 500 where n and m are the number of rows and columns in matrix
Example 1
Input

1
2
3
4
5
matrix = [
    [1, 3, 5],
    [0, 4, 6],
    [2, 2, 9]
]
Output
6

Explanation
The longest path is [0, 1, 3, 5, 6, 9]

Due to the strictly increasing condition, there cannot be any cycles
Question yourself on what will be the longest increasing path if I start from each cell (consider each cell as a starting point and traverse)
Keep a track of the longest increasing path from each cell. Compute this value once and use it whenever required.
"""
from functools import lru_cache
def longestIncreasingPath(nums):
    rows=len(nums)
    if not rows:
        return 0
    cols=len(nums[0])
    if not cols:
        return 0
   
    @lru_cache
    def f(r,c):
        ans=1
        for dr,dc in ((1,0),(0,1),(-1,0),(0,-1)):
            nr=r+dr
            nc=c+dc
            if 0<=nr<rows and 0<=nc<cols and nums[r][c]<nums[nr][nc]:
                ans=max(ans,1+f(nr,nc))
                
        return ans
    
    ans=1
    for r in range(rows):
        for c in range(cols):
            ans=max(ans,f(r,c))
    return ans
matrix = [
    [1, 3, 5],
    [0, 4, 6],
    [2, 2, 9]
]
print(longestIncreasingPath(matrix))

File: DPCodes/452-longestIncreasingPath.cpp
#include<iostream>
#include<vector>

using MATRIX=std::vector<std::vector<int>>;


int f(int r,int c,int R,int C,const MATRIX& m)
{
    int ans=1;
    MATRIX D{{1,0},{-1,0},{0,1},{0,-1}};
    for(auto &v:D)
    {
        int nr=v[0]+r;
        int nc=v[1]+c;
        if(nr>=0 &&nr<R &&nc>=0 &&nc<C&&(m[r][c]<m[nr][nc]))
        {
            ans=std::max(ans,f(nr,nc,R,C,m)+1);
        }
    }
    return ans;
}
int longestIncreasingPath(const MATRIX& m)
{
    int R=m.size(), C=m[0].size();
    if(R==0 || C==0)
        return 0;
    int ans=1;
    for(int r=0;r<R;++r)
    {
        for(int c=0;c<C;++c)
        {
            ans=std::max(ans,f(r,c,R,C,m));
        }
    }
    return ans;
}

int main()
{
    MATRIX m {
    {1, 3, 5},
    {0, 4, 6},
    {2, 2, 9}
    };
    std::cout<<longestIncreasingPath(m);
}